pytential/linalg/skeletonization.py

```diff
@@ -0,0 +1,790 @@
+    skl_ranges = np.zeros(blkindices.nblocks + 1, dtype=np.int32)
+
+    tgtindices = blkindices.row
+    srcindices = blkindices.col
+
+    from pytential.linalg import interp_decomp
+    for i in range(blkindices.nblocks):
+        k = id_rank
+
+        assert np.any(np.isfinite(tgt_mat[i, i])), \
+                np.where(np.isfinite(tgt_mat[i, i]))
+        assert np.any(np.isfinite(src_mat[i, i])), \
+                np.where(np.isfinite(src_mat[i, i]))
+
+        # skeletonize target points
+        k, idx, interp = interp_decomp(tgt_mat[i, i].T, k, id_eps)
+        assert k > 0
+
+        L[i, i] = interp.T
+        tgt_skl_indices[i] = tgtindices.block_indices(i)[idx[:k]]
+
+        # skeletonize source points
+        k, idx, interp = interp_decomp(src_mat[i, i], k, id_eps)
+        assert k > 0
+
+        R[i, i] = interp
+        src_skl_indices[i] = srcindices.block_indices(i)[idx[:k]]
+
+        skl_ranges[i + 1] = skl_ranges[i] + k
+        assert R[i, i].shape == (k, src_mat[i, i].shape[1])
+        assert L[i, i].shape == (tgt_mat[i, i].shape[0], k)
+
+    from pytential.linalg import make_block_index_from_array
+    src_skl_indices = make_block_index_from_array(
+            np.hstack(src_skl_indices), skl_ranges)
+    tgt_skl_indices = make_block_index_from_array(
+            np.hstack(tgt_skl_indices), skl_ranges)
+    skl_indices = MatrixBlockIndexRanges(tgt_skl_indices, src_skl_indices)
+
+    return L, R, skl_indices, src_mat, tgt_mat
+
+
+def skeletonize_block_by_proxy(
+        actx: PyOpenCLArrayContext,
+        ibcol: int, ibrow: int,
+        places: GeometryCollection,
+        proxy_generator: ProxyGenerator,
+        wrangler: SkeletonizationWrangler,
+        blkindices: MatrixBlockIndexRanges, *,
+        id_eps: Optional[float] = None,
+        id_rank: Optional[int] = None,
+        max_particles_in_box: Optional[int] = None
+        ) -> Tuple[np.ndarray, np.ndarray, MatrixBlockIndexRanges]:
+    L, R, sklindices, _, _ = _skeletonize_block_by_proxy_with_mats(actx,
+            ibcol, ibrow, places, proxy_generator, wrangler, blkindices,
+            id_eps=id_eps,
+            id_rank=id_rank,
+            max_particles_in_box=max_particles_in_box)
+
+    return L, R, sklindices
+
+# }}}
+
+
+# {{{ skeletonize_by_proxy
+
+@dataclass(frozen=True)
+class SkeletonizeResult:
+    r"""Result of a skeletonization procedure.
+
+    A matrix :math:`A` can be reconstructed using:
+
+    .. math::
+
+        A \approx L S R
+
+    where :math:`S = A_{I, J}` for a subset :math:`I` and :math:`J` of the
+    rows and columns of :math:`A`, respectively. This applies to each block
+    in :attr:`srcindices`.
+
+    .. attribute:: nblocks
+
+        Number of blocks that have been skeletonized.
+
+    .. attribute:: L
+
+        A block diagonal :class:`~numpy.ndarray` object array.
+
+    .. attribute:: R
+
+        A block diagonal :class:`~numpy.ndarray` object array.
+
+    .. attribute:: srcindices
```

Comment:
```suggestion
    .. attribute:: orig_indices
```
```suggestion
    .. attribute:: unskel_indices
```
?

Reply:
I renamed these to `tgt_src_index` and `skel_tgt_src_index`. Seems clear enough which ones are the skeletonized and which aren't to me. Thoughts?